File: src/client/client_session_manager.rs
use std::collections::HashMap;
use std::error::Error;
use std::fmt::format;
use std::io;
use std::io::ErrorKind;
use std::sync::Arc;
use lazy_static::lazy_static;
use tokio::net::TcpStream;
use tokio::sync::{Mutex, RwLock};
use crate::client::client_session::ClientSession;
use crate::client::{client_session, header_util};
use crate::dao::client_dao;
use crate::dao::dto::client_dto::ClientDto;
// use crate::client::{client_session, header_util};
// use crate::client::client_session::ClientSession;
// use crate::dao::client_dao;
// use crate::dao::dto::client_dto::ClientDto;

lazy_static! {

/**
 * 客户端ID对应的Socket连接
 */
    pub static ref clientSessionMap: RwLock<HashMap<i64, Arc<ClientSession>>> = RwLock::new(HashMap::new());
}

/**
 * 添加互斥锁
 */
// private val clientSessionMapLock = Mutex()

/**
 * 获取与客户端的会话
 */
pub async fn get_session(client_id: i64) -> Arc<ClientSession> {
    let map = clientSessionMap.read().await;
    let dto = map.get(&client_id).unwrap();
    dto.clone()
}

/**
 * 当前客户端数量
 */
pub async fn size() -> usize {
    clientSessionMap.read().await.len()
}

/**
 * 获取当前客户端会话列表
 */
// pub async fn getSessionList() -> List<ClientSession> {
//     return clientSessionMap.toList();
// }

/**
 * 获取当前客户端列表
 */
// pub async fn getClientList() -> List<ClientDto> {
//     return clientSessionMap.toList();
// }

/**
 * 添加客户端连接
 */
pub async fn validate(tcp: TcpStream) -> Result<(), Box<dyn Error>> {

    //得到头部数据
    let (tcp,head) = header_util::get_header(tcp).await?;
    let heads: Vec<_> = head.split("|").collect();

    //得到客户端key
    let key = heads[0];
    let client_opt = client_dao::selectByKey(key);
    if client_opt.is_none() {
        Err::<ErrorKind, io::Error>(io::Error::new(ErrorKind::Other, "key".to_owned() + &key + "不存在"));
        // clientSocket.close();
    }
    let client = client_opt.unwrap();
    if client.enable_state == 0 {
        Err::<ErrorKind, io::Error>(io::Error::new(ErrorKind::Other, "key".to_owned() + &key + "的客户端已停止服务"));
    }

    let client_id = client.id;

    //客户端ip
    // let ip = clientSocket.inetAddress.hostAddress;
    let ip = "";

    //从头部信息中得到客户端版本号
    let version = heads[1];

    client_dao::setClientInfo(client_id, ip.to_string(), version.to_string());

    //将客户端ID返回给客户端
    // sendClientId(client_id);


    holdOnClient(tcp,client);
    //
    // //将加密秘钥发送到客户端
    // sendClientSecurityKey(client.id);
    //
    // //开启该客户端下所有隧道监听
    // ProxyAcceptManager.accept(client);

    Ok(())
}

/**
 * 保持客户端连接
 */
async fn holdOnClient(tcp: TcpStream,client: ClientDto) {
    let client_id = client.id;

    //先移除之前的连接
    close(client_id);
    let session = ClientSession {
        client,
        clientSocket: Mutex::new(tcp),
        lastHeartBeatTime: 0,
    };
    {
        //将会话添加到map
        clientSessionMap.write().await.insert(client_id, Arc::new(session));
    }
    let session = clientSessionMap.read().await.get(&client_id).unwrap();
    client_session::start(session);
}

/**
 * 将客户端ID返回给客户端
 */
async fn sendClientId(clientID: i64) {
    send(clientID, header_util::SERVER_TO_CLIENT_ID as i8, clientID.to_string()).await;
}

// /**
//  * 将加密秘钥发送到客户端
//  */
// async fn sendClientSecurityKey(clientID: u64) {
//     this.clientSessionMap[clientID]?.send(SecurityUtil.clientKeyArray);
// }
//
// /**
//  * 向客户端申请TCP连接池请求
//  * @param clientID 客户端ID
//  * @param count 申请数量
//  */
// pub async fn sendTCPPoolRequest(clientID: Int, count: u8) {
//     send(clientID, HeaderUtil.SERVER_TCP_POOL_REQUEST, count.toString());
// }
//
// /**
//  * 向客户端申请UDP连接池请求
//  * @param clientID 客户端ID
//  * @param count 申请数量
//  */
// pub async fn sendUDPPoolRequest(clientID: Int, count: Int) {
//     send(clientID, HeaderUtil.SERVER_UDP_POOL_REQUEST, count.toString())
// }

/**
 * 往客户端发送数据
 * @param clientID 客户端ID
 * @param flag 头部标记
 * @param message 头部消息
 */
pub async fn send(clientID: i64, flag: i8, message: String) {
    clientSessionMap.read().await.get(&clientID).unwrap().sendHeader(flag, message).await;
}

/**
 * 关闭客户端
 */
pub async fn close(clientId: i64) {
    let mut map = clientSessionMap.read().await;
    if map.contains_key(&clientId) {
        let session = map.get(&clientId);
        session.unwrap().close().await;
        map.remove(&clientId);
    }
}

File: client/ClientSessionManager/ClientSessionManager.go
package ClientSessionManager

import (
	"DairoNPS/client/ClientSession"
	"DairoNPS/client/HeaderUtil"
	"DairoNPS/dao/ClientDao"
	"DairoNPS/dao/dto"
	"DairoNPS/pool/TCPPoolManager"
	"DairoNPS/proxy/ProxyAcceptManager"
	"log"
	"net"
	"strconv"
	"strings"
	"sync"
)

type ClientSessionManager struct{}

//往客户端发送指令的专用连接

/**
 * 客户端ID对应的Socket连接
 */
var clientSessionMap = make(map[int]*ClientSession.ClientSession)

/**
 * 添加互斥锁
 */
var clientSessionMapLock sync.Mutex

/**
 * 获取与客户端的会话
 */
//func getSession(clientID: Int) = this.clientSessionMap[clientID]

/**
 * 当前客户端数量
 */
//val size: Int
//    get() = this.clientSessionMap.size

/**
 * 获取当前客户端会话列表
 */
//fun getSessionList(): List<ClientSession> = this.clientSessionMapLock.withLock {
//    return this.clientSessionMap.map {
//        it.value
//    }
//}

/**
 * 获取当前客户端列表
 */
//suspend fun getClientList(): List<ClientDto> = this.clientSessionMapLock.withLock {
//    return this.clientSessionMap.map {
//        it.value.client
//    }
//}

/**
 * 添加客户端连接
 */
func Validate(clientSocket net.Conn) {

	//得到头部数据
	header := HeaderUtil.GetHeader(clientSocket)
	if len(header) == 0 {
		return
	}
	headers := strings.Split(header, "|")

	//得到客户端key
	key := headers[0]
	client := ClientDao.SelectByKey(key)
	if client == nil {
		log.Printf("-->客户端：%s不存在\n", key)
		clientSocket.Close()
		return
	}
	if client.EnableState == 0 {
		log.Printf("-->客户端：%s的客户端已被停止服务\n", key)
		clientSocket.Close()
		return
	}
	holdOnClient(client, clientSocket)

	////客户端ip
	//ip := clientSocket.RemoteAddr().String()
	//
	////从头部信息中得到客户端版本号
	//version := headers[0]
	//loginClientDto := dto.ClientDto{
	//	Id:      client.Id,
	//	Ip:      ip,
	//	Version: version,
	//}
	//ClientDao.SetClientInfo(loginClientDto)

	//开启该客户端下所有隧道监听
	ProxyAcceptManager.Accept(client)
}

/**
 * 保持客户端连接
 */
func holdOnClient(client *dto.ClientDto, clientSocket net.Conn) {
	//先移除之前的连接
	Close(client.Id)
	clientSessionMapLock.Lock()
	session := &ClientSession.ClientSession{
		Client:       client,
		ClientSocket: clientSocket,
	}
	clientSessionMap[client.Id] = session

	//初始化客户端连接池
	TCPPoolManager.InitEmptyPoolByClient(client.Id)
	go session.Start(Close)
	clientSessionMapLock.Unlock()
}

/**
 * 向客户端申请TCP连接池请求
 * @param clientId 客户端ID
 * @param count 申请数量
 */
func (csm *ClientSessionManager) SendTCPPoolRequest(clientId int, count int) {
	send(clientId, HeaderUtil.SERVER_TCP_POOL_REQUEST, strconv.Itoa(count))
}

/**
 * 向客户端申请UDP连接池请求
 * @param clientID 客户端ID
 * @param count 申请数量
 */
//suspend fun sendUDPPoolRequest(clientID: Int, count: Int) {
//    this.send(clientID, HeaderUtil.SERVER_UDP_POOL_REQUEST, count.toString())
//}

/**
 * 往客户端发送数据
 * @param clientID 客户端ID
 * @param flag 头部标记
 * @param message 头部消息
 */
func send(clientId int, flag byte, message string) {
	session := clientSessionMap[clientId]
	session.SendHead(flag, message)
}

/**
 * 关闭客户端
 */
func Close(clientId int) {
	clientSessionMapLock.Lock()
	if _, ok := clientSessionMap[clientId]; ok { //如果存在
		clientSession := clientSessionMap[clientId]

		//去关闭连接
		clientSession.Close()
		ClientDao.SetDataLen(clientSession.Client)
		delete(clientSessionMap, clientId)
	}
	clientSessionMapLock.Unlock()
}
